File: llist.cpp
#include "llist.h"
#include <iostream>

void CreateLList(LList* llist) {
    llist -> size = 0;
    llist -> head = nullptr;
    llist -> tail = nullptr;
}

LNode* CreateLNode(string data) {
    LNode* newNode = new LNode;
    newNode -> data = data;
    newNode -> pNext = nullptr;
    newNode -> pPrev = nullptr;
    return newNode;
}

void LPushHead(LList* llist, string data) {
    LNode* newNode = CreateLNode(data);
    if (!llist -> head) {
        llist -> head = llist -> tail = newNode;
    } else {
        newNode -> pNext = llist -> head;
        llist -> head -> pPrev = newNode;
        llist -> head = newNode;
    }
    llist -> size++;
}

void LPushTail(LList* llist, string data) {
    LNode* newNode = CreateLNode(data);
    if (!llist -> head) {
        llist -> head = llist -> tail = newNode;
    } else {
        newNode -> pPrev = llist -> tail;
        llist -> tail -> pNext = newNode;
        llist -> tail = newNode;
    }
    llist -> size++;
}

void LPushBefore(LList* llist, string data, string element) {
    if (llist -> head -> data == element) {
        LPushHead(llist, data);
        return;
    }
    LNode* current = llist -> head;
    while (current && current -> data != element) {
        current = current -> pNext;
    }
    if (current) {
        LNode* newNode = CreateLNode(data);
        newNode -> pNext = current;
        newNode -> pPrev = current -> pPrev;
        if (current -> pPrev) current -> pPrev -> pNext = newNode;
        current -> pPrev = newNode;
        llist -> size++;
    }
}

void LPushAfter(LList* llist, string data, string element) {
    if (!llist -> head) return;
    if (llist -> tail -> data == element) {
        LPushTail(llist, data);
        return;
    }
    LNode* current = llist -> head;
    while (current && current -> data != element) {
        current = current -> pNext;
    }
    if (current) {
        LNode* newNode = CreateLNode(data);
        newNode -> pPrev = current;
        newNode -> pNext = current -> pNext;
        if (current -> pNext) current -> pNext -> pPrev = newNode;
        current -> pNext = newNode;
        llist -> size++;
    }
}

void LPopHead(LList* llist) {
    if (!llist -> head) return;
    LNode* temp = llist -> head;
    llist -> head = llist -> head -> pNext;
    if (llist -> head) llist -> head -> pPrev = nullptr;
    else llist -> tail = nullptr;
    delete temp;
    llist -> size--;
}

void LPopTail(LList* llist) {
    if (!llist -> head) return;
    if (!llist -> tail -> pPrev) {
        delete llist -> head;
        llist -> head = llist -> tail = nullptr;
        llist -> size--;
        return;
    }
    LNode* temp = llist -> tail;
    llist -> tail = llist -> tail -> pPrev;
    llist -> tail -> pNext = nullptr;
    delete temp;
    llist -> size--;
}

void LPopElement(LList* llist, string element) {
    if (!llist -> head) return;
    LNode* current = llist -> head;
    while (current && current -> data != element) {
        current = current -> pNext;
    }
    if (current) {
        if (!current -> pPrev) LPopHead(llist);
        else if (!current -> pNext) LPopTail(llist);
        else {
            current -> pPrev -> pNext = current -> pNext;
            current -> pNext -> pPrev = current -> pPrev;
            delete current;
            llist -> size--;
        }
    }
}

bool LSearch(LList* llist, string element) {
    LNode* current = llist -> head;
    while (current) {
        if (current -> data == element) return true;
        current = current -> pNext;
    }
    return false;
}

void PrintLList(LList* llist) {
    cout << "[";
    LNode* current = llist -> head;
    while (current) {
        cout << current -> data;
        if (current -> pNext) cout << ", ";
        current = current -> pNext;
    }
    cout << "]" << endl;
}

void DestroyLList(LList* llist) {
    LNode* current = llist -> head;
    while (current) {
        LNode* temp = current;
        current = current -> pNext;
        delete temp;
    }
    llist -> head = llist -> tail = nullptr;
}

File: llist.go
package main

import "fmt"

type LNode struct {
	Data  string
	Next  *LNode
	Prev  *LNode
}

type LList struct {
	Size int
	Head *LNode
	Tail *LNode
}

func CreateLList() *LList {
	return &LList{
		Size: 0,
		Head: nil,
		Tail: nil,
	}
}

func CreateLNode(data string) *LNode {
	return &LNode{
		Data: data,
		Next: nil,
		Prev: nil,
	}
}

func (llist *LList) PushHead(data string) {
	newNode := CreateLNode(data)
	if llist.Head == nil {
		llist.Head = newNode
		llist.Tail = newNode
	} else {
		newNode.Next = llist.Head
		llist.Head.Prev = newNode
		llist.Head = newNode
	}
	llist.Size++
}

func (llist *LList) PushTail(data string) {
	newNode := CreateLNode(data)
	if llist.Head == nil {
		llist.Head = newNode
		llist.Tail = newNode
	} else {
		newNode.Prev = llist.Tail
		llist.Tail.Next = newNode
		llist.Tail = newNode
	}
	llist.Size++
}

func (llist *LList) PushBefore(data, element string) {
	if llist.Head == nil {
		return
	}
	if llist.Head.Data == element {
		llist.PushHead(data)
		return
	}
	current := llist.Head
	for current != nil && current.Data != element {
		current = current.Next
	}
	if current != nil {
		newNode := CreateLNode(data)
		newNode.Next = current
		newNode.Prev = current.Prev
		if current.Prev != nil {
			current.Prev.Next = newNode
		}
		current.Prev = newNode
		llist.Size++
	}
}

func (llist *LList) PushAfter(data, element string) {
	if llist.Head == nil {
		return
	}
	if llist.Tail.Data == element {
		llist.PushTail(data)
		return
	}
	current := llist.Head
	for current != nil && current.Data != element {
		current = current.Next
	}
	if current != nil {
		newNode := CreateLNode(data)
		newNode.Prev = current
		newNode.Next = current.Next
		if current.Next != nil {
			current.Next.Prev = newNode
		}
		current.Next = newNode
		llist.Size++
	}
}

func (llist *LList) PopHead() {
	if llist.Head == nil {
		return
	}
	llist.Head = llist.Head.Next
	if llist.Head != nil {
		llist.Head.Prev = nil
	} else {
		llist.Tail = nil
	}
	llist.Size--
}

func (llist *LList) PopTail() {
	if llist.Head == nil {
		return
	}
	if llist.Tail.Prev == nil {
		llist.Head = nil
		llist.Tail = nil
		llist.Size--
		return
	}
	llist.Tail = llist.Tail.Prev
	llist.Tail.Next = nil
	llist.Size--
}

func (llist *LList) PopElement(element string) {
	if llist.Head == nil {
		return
	}
	current := llist.Head
	for current != nil && current.Data != element {
		current = current.Next
	}
	if current != nil {
		if current.Prev == nil {
			llist.PopHead()
		} else if current.Next == nil {
			llist.PopTail()
		} else {
			current.Prev.Next = current.Next
			current.Next.Prev = current.Prev
			llist.Size--
		}
	}
}

func (llist *LList) Search(element string) bool {
	current := llist.Head
	for current != nil {
		if current.Data == element {
			return true
		}
		current = current.Next
	}
	return false
}

func (llist *LList) Print() {
	fmt.Print("[")
	current := llist.Head
	for current != nil {
		fmt.Print(current.Data)
		if current.Next != nil {
			fmt.Print(", ")
		}
		current = current.Next
	}
	fmt.Println("]")
}